Add version to R Syntax

boutinb opened the pull request:
Also 'data' and 'version' should be forbidden as option names.
Better handling of the option name, so that they don't conflict with each other, a formula name, or a reserved word.

boutinb committed:
Add version to R Syntax

Also 'data' and 'version' should be forbidden as option names.

JorisGoosen commented on QMLComponents/controls/jaspcontrol.cpp:
```diff
@@ -203,10 +206,22 @@ void JASPControl::componentComplete()
 		_form->addControl(this);
 	else
 	{
+		// The control is created dynamically, this is the case for row components.
```
🙇

JorisGoosen commented on QMLComponents/rsyntax/rsyntax.cpp:
```diff
@@ -155,7 +158,8 @@ QString RSyntax::generateWrapper() const
 	result += _form->name() + "Wrapper <- function(\n";
-	result += FunctionOptionIndent + "data = NULL";
+	result += FunctionOptionIndent + "data = NULL\n";
+	result += FunctionOptionIndent + "version = \"" + tq(AppInfo::version.asString()) + "\"";
```
This is wrong, you need to use the version of the module

JorisGoosen commented on QMLComponents/rsyntax/rsyntax.cpp:
```diff
@@ -85,7 +86,9 @@ QString RSyntax::generateSyntax() const
 	result = _analysisFullName(true) + "(\n";
-	result += FunctionOptionIndent + "data = NULL";
+	result += FunctionOptionIndent + "data = NULL\n";
+	result += FunctionOptionIndent + "version = \"" + tq(AppInfo::version.asString()) + "\"";
```
as you should here

boutinb committed:
Use analysis version in place of JASP version

JorisGoosen commented on QMLComponents/analysisbase.h:
```diff
@@ -27,6 +28,7 @@ class AnalysisBase : public QObject
 	virtual const std::string & module()									const	{ return emptyString;		}
 	virtual const std::string & name()										const	{ return emptyString;		}
 	virtual const std::string & title()										const	{ return emptyString;		}
+	virtual const Version & version()										const	{ return AppInfo::version;	}
```
Doesnt this mean that the versionnumbers in R-only mode will be totally wrong?

JorisGoosen replied:
As in, with the current code all the stuff that has only access to AnalysisBase will always return whatever was the version of JASP when jaspBase gets compiled right?

Why not just add `_version` in AnalysisBase and not have an override in Analysis and then just make sure a value gets set into it.

boutinb committed:
Set version in AnalysisBase

JorisGoosen reviewed (changes requested):
This is better... Except for a crucial problem.

The moduleversion is stored in `_moduleVersion` and the JASP version is `_version` .
_version isnt really used in JASP anymore I think.

JorisGoosen commented:
In fact, "version" on Analysis is there basically only to load jaspfiles from pre 0.15 jasps

JorisGoosen commented:
Maybe add that as a comment to the definition of `_version` in Analysis (as there is no reason to have the jasp-version specified when running in pure R anyway)

boutinb committed:
Confusion with version, oldVersion and moduleVersion

. oldVersion is used for pre-upgrader version: rename it preUpgraderVersion
. version is obsolete and can be removed
. moduleVersion is set in the analysisBase class

boutinb committed:
Avoid crash

JorisGoosen commented on Desktop/analysis/analysis.cpp:
```diff
@@ -453,9 +448,9 @@ void Analysis::setStatus(Analysis::Status status)
+		_wasUpgraded = false;
 
-		_moduleVersion = _dynamicModule->version();
+		_moduleVersion = _dynamicModule ?  _dynamicModule->version() : AppInfo::version;
```
Why are you changing the behaviour?

JorisGoosen replied:
this will all break loading old old files won't it?

JorisGoosen replied:
```suggestion
	_moduleVersion = _dynamicModule->version();
```

boutinb replied:
Apparently for old JASP file, the _dynamicModule can be null, so it is here only to make it not crash, maybe I can then better set the  moduleVersion to 0.

JorisGoosen replied:
<img width="1241" alt="image" src="https://user-images.githubusercontent.com/29062713/205967754-c018ed7f-fae6-494f-a012-6ec9b93219c1.png">

the right diff is from a log from your branch, it just loaded the attached file. 
[oudJaspFile-johnnyRfClassification.zip](https://github.com/jasp-stats/jasp-desktop/files/10168327/oudJaspFile-johnnyRfClassification.zip)

It should not have changed the version yet, as it hasn't been refreshed. I think this might cause the errormessage to dissappear

JorisGoosen replied:
First of all the upgrader handles `dynamicModule` being null, see: 
```
void DynamicModule::applyUpgrade(const std::string	& function,	const Version & version, Json::Value & analysesJson, UpgradeMsgs & msgs, StepsTaken & stepsTaken)
{
	if(!analysesJson.isMember("dynamicModule"))
		analysesJson["dynamicModule"] = asJsonForJaspFile(function);
	
	_upgrades->applyUpgrade(function, version, analysesJson, msgs, stepsTaken);
}
```

But it needs to have the information available to do this as seen at:
https://github.com/jasp-stats/jasp-desktop/blob/4bed733e759f11b14b19d0aa1f992221f21e07b9/Desktop/modules/upgrader/upgrader.cpp#L122

JorisGoosen commented on Desktop/analysis/analysis.cpp:
```diff
@@ -394,7 +390,6 @@ Json::Value Analysis::asJSON(bool withRSource) const
 	analysisAsJson["rfile"]			= _rfile;
 	analysisAsJson["hasReport"]		= _hasReport;
 	analysisAsJson["progress"]		= _progress;
-	analysisAsJson["version"]		= _version.asString();
```
Its still useful to know which version of JASP was used to make an analysis right?

JorisGoosen replied:
Especially as the dynamic module might be, you know, dynamic

boutinb replied:
The _version is an obsolete property. Only _moduleVersion is used, and it is stored with the "dyamicModule" part of the json value (see Analyses::createFromJaspFileEntry)

JorisGoosen replied:
I guess we could rely on the `version` in `Analysis` for the JASP version.

JorisGoosen commented on Desktop/analysis/analysis.cpp:
```diff
@@ -394,7 +390,6 @@ Json::Value Analysis::asJSON(bool withRSource) const
 	analysisAsJson["rfile"]			= _rfile;
 	analysisAsJson["hasReport"]		= _hasReport;
 	analysisAsJson["progress"]		= _progress;
```
```suggestion
	analysisAsJson["progress"]		= _progress;
	analysisAsJson["version"]		= _version.asString();
```

JorisGoosen commented:
I think some of those changes you made with regards to needsRefresh might have caused this bug:
<img width="1333" alt="image" src="https://user-images.githubusercontent.com/29062713/205966088-0e3eadd4-e58d-4464-b6c1-d6e81f03a20a.png">

Left is with your code and right is 0.16.4, file below:

[oudJaspFile-bLinReg.zip](https://github.com/jasp-stats/jasp-desktop/files/10168279/oudJaspFile-bLinReg.zip)

JorisGoosen commented:
So if you could change this PR to *not change the behaviour* that would be lovely.

~So keeping "version" in there so we can track the original version of JASP used.~
And keep the `setVersion` function on Analysis, or was there a reason you removed it?

Cause I think it is implicated in getting the error message to show, but am not sure

JorisGoosen commented:
Hmm, it seems the problem was already there in `stable`:
<img width="1544" alt="image" src="https://user-images.githubusercontent.com/29062713/206182072-e450b59c-0794-4c80-aad3-551e266c3684.png">

Ill check your code again later.
